src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/SemanticTokensEditsDiffer.cs

```diff
@@ -67,37 +67,59 @@ private Container<SemanticTokensEdit> ProcessEdits(IReadOnlyList<DiffEdit> diffs
                         }
                         else
                         {
-                            results.Add(new SemanticTokensEdit
+                            razorResults.Add(new RazorSemanticTokensEdit
                             {
                                 Start = diff.Position,
-                                Data = ImmutableArray<int>.Empty,
                                 DeleteCount = 1,
                             });
                         }
                         break;
                     case DiffEdit.Type.Insert:
                         if (current != null &&
-                            current.Data.HasValue &&
-                            current.Data.Value.Any() &&
+                            current.Data != null &&
+                            current.Data.Any() &&
                             current.Start == diff.Position)
                         {
-                            current.Data = current.Data.Append(NewArray[diff.NewTextPosition!.Value]).ToImmutableArray();
+                            current.Data.Add(NewArray[diff.NewTextPosition!.Value]);
                         }
                         else
                         {
-                            var semanticTokensEdit = new SemanticTokensEdit
+                            var semanticTokensEdit = new RazorSemanticTokensEdit
                             {
                                 Start = diff.Position,
-                                Data = ImmutableArray.Create(NewArray[diff.NewTextPosition!.Value]),
+                                Data = new List<int>{
+                                    NewArray[diff.NewTextPosition!.Value],
+                                },
                                 DeleteCount = 0,
                             };
-                            results.Add(semanticTokensEdit);
+                            razorResults.Add(semanticTokensEdit);
                         }
                         break;
                 }
             }
 
-            return results;
+            var results = razorResults.Select(e => e.ToSemanticTokensEdit());
+            return results.ToList();
+        }
+
+        // We need to have a shim class because SemanticTokensEdit.Data is Immutable, so if we operate on it directly then every time we append an element we're allocating an entire new array.
+        // In some large (but not implausibly so) copy-paste scenarios that can cause long delays and large allocations.
+        private class RazorSemanticTokensEdit
+        {
+            public int Start { get; set; }
+            public int DeleteCount { get; set; }
+            public IList<int>? Data { get; set; } = new List<int>();
```

Comment:
Instead of creating a new list here if you did it lazily would you be able to reduce allocations even more? I'm assuming that there are occasions when `Data.Count == 0`